pkg/schedule/schedulers/balance_keyrange.go

```diff
@@ -0,0 +1,488 @@
+// Copyright 2024 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"bytes"
+	"cmp"
+	"slices"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/log"
+
+	"github.com/tikv/pd/pkg/core"
+	"github.com/tikv/pd/pkg/errs"
+	sche "github.com/tikv/pd/pkg/schedule/core"
+	"github.com/tikv/pd/pkg/schedule/operator"
+	"github.com/tikv/pd/pkg/schedule/plan"
+	"github.com/tikv/pd/pkg/schedule/types"
+)
+
+type storeRegionSet struct {
+	ID   uint64
+	Info *metapb.Store
+	// If the region still exists durgin migration
+	RegionIDSet  map[uint64]bool
+	OriginalPeer map[uint64]*metapb.Peer
+}
+
+type migrationOp struct {
+	FromStore    uint64          `json:"from_store"`
+	ToStore      uint64          `json:"to_store"`
+	ToStoreInfo  *metapb.Store   `json:"to_store_info"`
+	OriginalPeer *metapb.Peer    `json:"original_peer"`
+	Regions      map[uint64]bool `json:"regions"`
+}
+
+func pickRegions(n int, fromStore *storeRegionSet, toStore *storeRegionSet) *migrationOp {
+	o := migrationOp{
+		FromStore:   fromStore.ID,
+		ToStore:     toStore.ID,
+		ToStoreInfo: toStore.Info,
+		Regions:     make(map[uint64]bool),
+	}
+
+	for range n {
+		for r, fromHasRegion := range fromStore.RegionIDSet {
+			if !fromHasRegion {
+				continue
+			}
+			if toHasRegion, exist := toStore.RegionIDSet[r]; !exist || !toHasRegion {
+				// If toStore doesn't has this region, then create a move op.
+				o.Regions[r] = false
+				o.OriginalPeer = fromStore.OriginalPeer[r]
+				toStore.RegionIDSet[r] = true
+				fromStore.RegionIDSet[r] = false
+				break
+			}
+		}
+	}
+	return &o
+}
+
+func buildMigrationPlan(stores []*storeRegionSet) ([]int, []int, []*migrationOp, int) {
```

Comment:
We also need to make store balance under constraint. `expectedCount` is not accurate.

Reply:
Then I will adopt the region-wise strategy.

Reply:
Does it means we may generate an operator that move some Region out of the given range to keep the "store balance"?